torch/_inductor/ir.py

```diff
@@ -4113,6 +4113,123 @@ def codegen(self, wrapper):
         wrapper.writeline(f"_register_tensor_work({output_name}, {output_name}_work)")
 
 
+class MultiOutputNoSizeAssert(MultiOutput):
+    """
+    Extract partial output from a multi-output OP.
+        Works like MultiOutput but doesn't assert size. This must be a property guaranteed by the op emiting this.
+    """
+
+    def codegen(self, wrapper):
+        wrapper.writeline(
+            f"{self.get_name()} = {self.inputs[0].get_name()}{self.index}"
+        )
+
+
+class ForceInPlace(ExternKernel):
+    """
+    Helper OP to encode an in/out argument that tries to make it inplace whenever possible.
+    Wrap the input of your inplace op to enable this behavior.
+
+    TODO: We should test whether wait_tensor can be a victim of reordering and lead to data races.
+    """
+
+    def codegen(self, wrapper):
+        input_name = self.inputs[0].codegen_reference()
+        output_name = self.get_name()
+        if not wrapper.did_reuse(self, self.inputs[0]):
+            wrapper.writeline(f"{output_name}.copy_({input_name}) #no reuse")
+
+    def __init__(self, layout, input):
+        input = self.realize_input(input)
+        super().__init__(None, layout, self.unwrap_storage([input]), ())
+        self.name = V.graph.register_buffer(self)
+
+    def should_allocate(self):
+        return True
+
+
+class AllReduceCoalesced(ExternKernel):
+    def __init__(self, layout, inputs, constant_args, reduce_op):
+        super().__init__(None, layout, inputs, constant_args)
+        self.reduce_op = reduce_op
+        self.name = V.graph.register_buffer(self)
+
+    def should_allocate(self):
+        return False
+
+    @classmethod
+    def create(
+        cls,
+        inputs: List["TensorBox"],
+        reduce_op: str,
+        tag: str,
+        ranks: List[int],
+        group_size: int,
+    ):
+        res = []
+
+        def wrap_input(var):
+            nonlocal res
+            op = ForceInPlace(
+                FlexibleLayout(var.get_device(), var.get_dtype(), var.get_size()), var
+            )
+            res.append(op)
```

Comment:
should `res` append a `ExternKernel` or a `TensorBox`? I think it should be `TensorBox` as that would be later passed into `AllreduceCoaleseced` inductor node?

Reply:
This is really tricky actually
The convention for lowering.py is to return IR nodes and not TensorBox from the `create` funs.
OTOH `AllreduceCoalesced` actually needs a list of `TensorBox` nodes.

This is cleared up in the next PR as I properly decouple inplace in/out nodes from the actual collectives. It took me a dozen iterations to arrive at that design and it's kinda ok.